components/lcd/esp_lcd_st7796/esp_lcd_st7796.c

```diff
@@ -1,353 +1,38 @@
 /*
- * SPDX-FileCopyrightText: 2023 Espressif Systems (Shanghai) CO LTD
+ * SPDX-FileCopyrightText: 2024 Espressif Systems (Shanghai) CO LTD
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#include <stdlib.h>
-#include <sys/cdefs.h>
-#include "freertos/FreeRTOS.h"
-#include "freertos/task.h"
-#include "esp_lcd_panel_interface.h"
-#include "esp_lcd_panel_io.h"
-#include "esp_lcd_panel_vendor.h"
-#include "esp_lcd_panel_ops.h"
-#include "esp_lcd_panel_commands.h"
-#include "driver/gpio.h"
-#include "esp_log.h"
+#include "soc/soc_caps.h"
 #include "esp_check.h"
+#include "esp_lcd_types.h"
 
 #include "esp_lcd_st7796.h"
+#include "esp_lcd_st7796_interface.h"
 
-static const char *TAG = "st7796";
+const char *TAG = "st7796";
 
-static esp_err_t panel_st7796_del(esp_lcd_panel_t *panel);
-static esp_err_t panel_st7796_reset(esp_lcd_panel_t *panel);
-static esp_err_t panel_st7796_init(esp_lcd_panel_t *panel);
-static esp_err_t panel_st7796_draw_bitmap(esp_lcd_panel_t *panel, int x_start, int y_start, int x_end, int y_end, const void *color_data);
-static esp_err_t panel_st7796_invert_color(esp_lcd_panel_t *panel, bool invert_color_data);
-static esp_err_t panel_st7796_mirror(esp_lcd_panel_t *panel, bool mirror_x, bool mirror_y);
-static esp_err_t panel_st7796_swap_xy(esp_lcd_panel_t *panel, bool swap_axes);
-static esp_err_t panel_st7796_set_gap(esp_lcd_panel_t *panel, int x_gap, int y_gap);
-static esp_err_t panel_st7796_disp_on_off(esp_lcd_panel_t *panel, bool off);
-
-typedef struct {
-    esp_lcd_panel_t base;
-    esp_lcd_panel_io_handle_t io;
-    int reset_gpio_num;
-    bool reset_level;
-    int x_gap;
-    int y_gap;
-    uint8_t fb_bits_per_pixel;
-    uint8_t madctl_val; // save current value of LCD_CMD_MADCTL register
-    uint8_t colmod_val; // save current value of LCD_CMD_COLMOD register
-    const st7796_lcd_init_cmd_t *init_cmds;
-    uint16_t init_cmds_size;
-} st7796_panel_t;
-
-esp_err_t esp_lcd_new_panel_st7796(const esp_lcd_panel_io_handle_t io, const esp_lcd_panel_dev_config_t *panel_dev_config, esp_lcd_panel_handle_t *ret_panel)
+esp_err_t esp_lcd_new_panel_st7796(const esp_lcd_panel_io_handle_t io, const esp_lcd_panel_dev_config_t *panel_dev_config,
+                                   esp_lcd_panel_handle_t *ret_panel)
 {
-    esp_err_t ret = ESP_OK;
-    st7796_panel_t *st7796 = NULL;
-    gpio_config_t io_conf = { 0 };
+    ESP_LOGI(TAG, "version: %d.%d.%d", ESP_LCD_ST7796_VER_MAJOR, ESP_LCD_ST7796_VER_MINOR, ESP_LCD_ST7796_VER_PATCH);
+    ESP_RETURN_ON_FALSE(panel_dev_config && ret_panel, ESP_ERR_INVALID_ARG, TAG, "Invalid arguments");
+    st7796_vendor_config_t *vendor_config = (st7796_vendor_config_t *)panel_dev_config->vendor_config;
+    ESP_RETURN_ON_FALSE(vendor_config, ESP_ERR_INVALID_ARG, TAG, "`vendor_config` is necessary");
```

Comment:
This is breaking change. Can we check it only when we need it? For I80/SPI we don't need `vendor_config`, right?

Reply:
Your point is indeed valid; I overlooked the fact that this is a breaking change. The issue has been fixed.